server/bundler/remove-parent-modules-monkeypatch.js

```diff
@@ -0,0 +1,116 @@
+// this is really gross.
+//
+// originally from
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+"use strict"; //eslint-disable-line
+
+var forEach = require( 'lodash/forEach' ),
+	filter = require( 'lodash/filter' ),
+	indexOf = require( 'lodash/indexOf' ),
+	pullAll = require( 'lodash/pullAll' ),
+	nanotimer = require( 'nanotimer' );
+
+var RemoveParentModulesPlugin = require( 'webpack/lib/optimize/RemoveParentModulesPlugin' );
+var Chunk = require( 'webpack/lib/Chunk' );
+var Module = require( 'webpack/lib/Module' );
+
+module.exports = RemoveParentModulesPlugin;
+
+function hasModule( chunk, module, checkedChunks ) {
+	if ( ~ indexOf( chunk.modules, module ) ) {
+		return [chunk];
+	}
+	if ( chunk.entry ) {
+		return false;
+	}
+	const uncheckParents = filter( chunk.parents, function( c ) {
+		return ! checkedChunks.has( c );
+	} );
+	return allHaveModule( uncheckParents, module, checkedChunks );
+}
+
+function allHaveModule( someChunks, module, checkedChunks ) {
+	if ( ! checkedChunks ) {
+		checkedChunks = new Set();
+	}
+	let chunks = new Set();
+	for ( let i = 0; i < someChunks.length; i++ ) {
+		checkedChunks.add( someChunks[i] );
+		const subChunks = hasModule( someChunks[i], module, checkedChunks );
+		if ( ! subChunks ) {
+			return false;
+		}
+		addToSet( chunks, subChunks );
+	}
+	return chunks;
+}
+
+function addToSet( set, items ) {
+	forEach( items, function( item ) {
+		set.add( item );
+	} );
+}
+
+Chunk.prototype._calypso_removeModules = function( modules ) {
+	this.modules = pullAll( this.modules, modules );
+	forEach( modules, function( module ) {
+		module._calypso_removeChunk( this );
+	} );
+}
+
+const oldAddParent = Chunk.prototype.addParent;
+Chunk.prototype.addParent = function( parent ) {
+	console.log( 'Adding %s as a parent of %s', parent.name, this.name );
+	return oldAddParent.apply( this, arguments );
+}
+
+Module.prototype._calypso_removeChunk = function( chunk ) {
+	this.chunks = pullAll( this.chunks, chunk );
+}
+
+RemoveParentModulesPlugin.prototype.apply = function( compiler ) {
+	compiler.plugin( 'compilation', function( compilation ) {
+		compilation.plugin( [ 'optimize-chunks', 'optimize-extracted-chunks' ], function( chunks ) {
+			console.time( 'remove-parents' );
+			/* So what are we doing here.
+			 * We're walking each chunks modules
+			 * If all of a chunk's parents have the module, remove it from the current chunks
+			 * If the chunk is an entry point, ignore it
+			 */
+			console.log( '%d chunks to process', chunks.length );
+			forEach( chunks, function( chunk ) {
+				console.time( '  remove-parents-' + chunk.name );
+				console.log( '    Processing %s: %d parents', chunk.name, chunk.parents.length );
+				chunk.parents.forEach( function( parent ) { console.log( '      %s', parent.name ); } );
+				let removals = 0, modulesToPull = [];
+				forEach( chunk.modules.slice(), function( module ) {
```

Comment:
i'm pretty sure we don't _have_ to slice-to-clone the module array. We can walk it backwards instead, then traversal is safe.